Various changes to <evaluate/>

bertfrees opened the pull request:
- Allow evaluation of `$volume` in volume transitions and in normal flow 
- Allow `<evaluate expression="$page"/>`, where you previously had to say `(round $page)` or `(format {0} $page)`
- Be more strict about what is allowed in an expression 
- Don't normalize space in evaluate expressions

bertfrees committed:
Allow evaluation of "$volume" in volume transitions and in normal flow

bertfrees committed:
Allow <evaluate expression="$page"/>

i.e. don't require the variable to be wrapped in a "(round ...)" or "(format {0} ...)"

Better would be to completely remove ExpressionTools and resolve variables in ExpressionImpl.

bertfrees committed:
Be more strict about what is allowed in an expression

For example, don't allow "$foo" when the variable "foo" is not declared.

bertfrees committed:
Don't normalize space in evaluate expressions

bertfrees commented:
I have rebased this onto master. See the corresponding commits in https://github.com/brailleapps/dotify.formatter.impl/compare/master...sbsdev:super/sbs.

PaulRambags commented on src/org/daisy/dotify/formatter/impl/search/SheetIdentity.java:
```diff
@@ -8,9 +8,14 @@
+			if (volumeIndex != null)
+				volumeIndex = null;
+			if (volumeGroup == null)
+				throw new IllegalArgumentException();
```
An exception is thrown without any explanation. Better would be:
`throw new IllegalArgumentException("Parameter volumeGroup is null");`

bertfrees replied:
I think that makes hardly any difference.

It should however be explained in the Javadoc of this constructor which arguments are valid and when a IllegalArgumentException is thrown.

kalaspuffar replied:
I agree, throwing an exception if handled correctly with the logging framework should give you a line number. But perhaps we could add other contextual information that could help us in the debugging efforts? For example 

```
throw new IllegalArgumentException("Parameter volumeGroup is null, this occurred on  sheet " + sheetIndex + " and the volume index was " +  volymeIndex);
```

Could be worded differently but perhaps that would help us try to identify the issue?

bertfrees replied:
I'm afraid even such information wouldn't make sense. These exceptions are more like assertions. If there is a bug somewhere that these conditions would catch, the code wouldn't even compile. Basically all the unit tests would fail. I think the signature of this and other methods that accept these variables could be improved so that the checks are not needed. E.g. instead of passing the `space` and `volumeGroup` variables around and checking them everywhere you can just create an object from them once and do the check once. Then all the methods that use it can be sure that the object is valid.

kalaspuffar commented on src/org/daisy/dotify/formatter/impl/page/PageSequenceBuilder2.java:
```diff
@@ -164,7 +170,6 @@ public PageImpl nextPage(int pageNumberOffset, boolean hyphenateLastLine, Option
 		for (String id : ret.getIdentifiers()) {
 			blockContext.getRefs().setPageNumber(id, ret.getPageNumber());
 		}
-		//This is for pre/post volume contents, where the volume number is known
```
Why did we remove this comment? Have we moved the pre/post volume contents or do we feel that the code is self-explanatory?

bertfrees replied:
I thought the comment was not accurate anymore, because the volume number is now always known, not only in pre/post volume contents but also in the volume body. However I just checked that the `CrossReferenceHandler#getVolumeNumber()` method is only used in `TocSequenceEventImpl` at the moment, and `toc-sequence` is currently only allowed in pre/post content, so technically the part that says "This is for pre/post volume contents" is still correct. But maybe the comment should be more specific and say that it is for use in TOCs.

kalaspuffar commented:
Hi @bertfrees 

I've prepared this PR to run regression tests as the changes we discuss is of minor nature. In order to merge this one though we need to rebase it and handle all the code style issues. I got 9 instances of missing curly braces when I tried to build this PR.

Best regards
Daniel

kalaspuffar commented on src/org/daisy/dotify/formatter/impl/VolumeProvider.java:
```diff
@@ -181,12 +181,13 @@ public double getCost(SplitPointDataSource<Sheet, ?> units, int index, int break
 
 		crh.setReadOnly();
 		SheetDataSource data = groups.currentGroup().getUnits();
+		data.setCurrentVolumeNumber(currentVolumeNumber);
```
Hi @bertfrees 

When we run the regression tests we have noticed that it generates a lot of changes with material moving around in the volumes. Different parts of articles moving to the previous volume in many cases. 

The volume amount doesn't change and most of the text is identical it's just the distribution of the text in volumes that differ. 

Why did we introduce this change?

Best regards
Daniel

kalaspuffar replied:
Hi @bertfrees 

I've now run the full test suite without this line and did not have any changes. I'll run it again with the line to see the actual impact of this line. > 6 books before I stopped it at 30 min.

Best regards
Daniel

kalaspuffar replied:
Hi again.

I've run the full suite now and we had 11 books with changes so not high impact but some changes. So if we feel that this change is required we could manually verify these changes.

Best regards
Daniel

bertfrees replied:
This is the essential part of the change which updates the `DefaultContext` of the current `SheetDataSource` (which updates the `BlockContext` in `PageSequenceBuilder2`, which updates the `BlockContext` in `RowGroupDataSource`) whenever a new volume is started, so that `$volume` can be resolved in expressions. That this influences the volume splitting is not expected. I would say that if we let this pass we're possibly missing a bug.

The only thing I can think of that could have influenced the volume splitting is if the behavior of `SheetIdentity#equals` would have changed, but this is not the case because of this line: https://github.com/brailleapps/dotify.formatter.impl/blob/d7e3eb780b81eaec625431dcb4d3fda543c7b2a8/src/org/daisy/dotify/formatter/impl/search/SheetIdentity.java#L13

Maybe I can try to debug with one of the failing books?

kalaspuffar replied:
Hi @bertfrees 

It did fix the issues with all books so no more regressions.

Best regards
Daniel

bertfrees commented:
> I got 9 instances of missing curly braces when I tried to build this PR.

Fixed now. And I rebased to branch onto master.

bertfrees committed:
Don't collect id-to-volume mapping until pages have come out of SheetDataSource

Initially, for pages in the volume body, the id-to-volume mapping was
done in VolumeProvider, but since the previous commit it was done also
in PageSequenceBuilder#nextPage, so in two places.

For pages in pre-/post-content, the id-to-volume mapping was only done
in PageSequenceBuilder#nextPage.

This commit moves the logic to one place, in VolumeProvider. Apart
from simplifying the code, this also fixes an issue in the regression
test suite. I have no proper explanation for it.

bertfrees committed:
In VolumeProvider align nextBodyContents and updateVolumeContents methods more

bertfrees commented:
@kalaspuffar I added commit 67da750 which fixes the regression in the book that you sent me. I hope it will fix the 10 other books too. Unfortunately I can't explain how the regression was caused and why my commit would have any effect. First I thought that it had something to do with pages that are produced (by `PageSequenceBuilder2`) but do not end up in the volume. But think this should not be the problem because of the `crh.setReadOnly()` and `crh.setReadWrite()` calls in `VolumeProvider#nextBodyContents` before/after the "find" phase, and in the "split" phase no pages are produced that will not end up in the volume.

kalaspuffar commented:
Hi @bertfrees 

Except for the Exceptions and the addition of better error messages, this is ready for merge. Are you going to make any improvements to mine or @PaulRambags suggestions or do we consider this done?

Best regards
Daniel
